components/brave_wallet/browser/solana_keyring.cc

```diff
@@ -57,4 +67,99 @@ std::string SolanaKeyring::GetAddressInternal(HDKeyBase* hd_key_base) const {
   return hd_key->GetBase58EncodedPublicKey();
 }
 
+// Create a valid program derived address without searching for a bump seed.
+// https://docs.rs/solana-program/latest/solana_program/pubkey/struct.Pubkey.html#method.create_program_address
+// static
+absl::optional<std::string> SolanaKeyring::CreateProgramDerivedAddress(
+    const std::vector<std::vector<uint8_t>>& seeds,
+    const std::string& program_id) {
+  const std::string pda_marker = "ProgramDerivedAddress";
+
+  std::vector<uint8_t> program_id_bytes;
+  if (!Base58Decode(program_id, &program_id_bytes, kSolanaPubkeySize))
+    return absl::nullopt;
+
+  if (seeds.size() > kMaxSeeds) {
+    return absl::nullopt;
+  }
+
+  for (const auto& seed : seeds) {
+    if (seed.size() > kMaxSeedLen) {
+      return absl::nullopt;
+    }
+  }
+
+  std::vector<uint8_t> buffer;
+  for (const auto& seed : seeds) {
+    buffer.insert(buffer.end(), seed.begin(), seed.end());
+  }
+
+  buffer.insert(buffer.end(), program_id_bytes.begin(), program_id_bytes.end());
+  buffer.insert(buffer.end(), pda_marker.begin(), pda_marker.end());
+
+  auto hash_array = crypto::SHA256Hash(buffer);
+  std::vector<uint8_t> hash_vec(hash_array.begin(), hash_array.end());
+
+  if (!bytes_are_curve25519_point(
+          rust::Slice<const uint8_t>{hash_vec.data(), hash_vec.size()})) {
+    return Base58Encode(hash_vec);
+  }
+
+  return absl::nullopt;
+}
+
+// Find a valid program derived address and its corresponding bump seed.
+// https://docs.rs/solana-program/latest/solana_program/pubkey/struct.Pubkey.html#method.find_program_address
+// static
+absl::optional<std::string> SolanaKeyring::FindProgramDerivedAddress(
+    const std::vector<std::vector<uint8_t>>& seeds,
+    const std::string& program_id,
+    uint8_t* ret_bump_seed) {
+  std::vector<std::vector<uint8_t>> seeds_with_bump(seeds);
+  std::vector<uint8_t> bump_seed = {UINT8_MAX};
+  while (bump_seed[0] != 0) {
+    seeds_with_bump.push_back(bump_seed);
+
+    auto address = CreateProgramDerivedAddress(seeds_with_bump, program_id);
+    if (address) {
+      if (ret_bump_seed)
+        *ret_bump_seed = bump_seed[0];
+
+      return address;
+    }
+
+    seeds_with_bump.pop_back();
+    bump_seed[0]--;
```

Comment:
++bump_seed[0]

Reply:
Fixed in https://github.com/brave/brave-core/pull/12722/commits/0d677d2b6fe0ea1b930bf272d5ca7e79c8359cdb